lib/charms/layer/jenkins/api.py

```diff
@@ -124,7 +134,12 @@ def _make_client(self):
         # TODO: also handle regenerated tokens
         if token is None:
             client = jenkins.Jenkins(self.url, user, creds.password())
-            token = client.run_script(GET_TOKEN_SCRIPT.format(user)).strip()
+            # If we're using Jenkins >= 2.129 we need to request a new token.
+            jenkins_version = Packages().jenkins_version()
```

Comment:
It'd be preferable to use dependency injection for the Packages() object, that'd mean to add something like:

```
def __init__(self, packages=None):
    self._packages = packages or Packages()
```

to the Api class, and then leverage it in tests by building the Api object using the fake packages implementation, e.g.:

```
 def setUp(self):
        super(ApiTest, self).setUp()
        self.useFixture(JenkinsConfiguredAdmin(self.fakes))
        self.fakes.jenkins.scripts[GET_TOKEN_SCRIPT.format("admin")] = "abc\n"
        self.apt = AptStub()
        self.packages = Packages(apt=self.apt)
        self.api = Api(packages=self.packages)

```

This would avoid the need to use @mock.patch (the tests that want the legacy version can just call ```self.apt.set_legacy_jenkins_version```, see the other comment below).

Not blocking comment, just a nit/hit, fix or leave it as you feel.